File: C++/CombinationSum.cpp
//Ian Hays
//12-22-20
//https://leetcode.com/problems/combination-sum/
//SC: O(N^2) TC: O(N^2)

class Solution {
public:
    vector<vector<int>> res;
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<int> combination;
        dfs(candidates, target, combination,0);
        return res;
    }
    
    void dfs(vector<int>& candidates, int target, vector<int>& combination, int start){
        if(target == 0){
            res.push_back(combination);
            return;
        }
        if(target < 0) return;
        for(int i = start; i < candidates.size(); i++){
            combination.push_back(candidates[i]);
            dfs(candidates, target-candidates[i], combination, i);
            combination.pop_back();
        }
    }
};

File: C#/CombinationSum.cs
//Ian Hays
//12-22-20
//https://leetcode.com/problems/combination-sum/
//SC: O(N^2) TC: O(N^2)

public class Solution {
    public IList<IList<int>> CombinationSum(int[] candidates, int target) {
        IList<IList<int>> result = new List<IList<int>>();
        List<int> combination = new List<int>();
        dfs(ref result, candidates, target, ref combination, 0);
        return result;
    }
    
    void dfs(ref IList<IList<int>> result, int[] candidates, int target, ref List<int> combination, int start){
        if(target == 0){
            result.Add(new List<int>(combination));
            return;
        }
        if(target < 0) return;
        for(int i = start; i < candidates.Length; i++){
            combination.Add(candidates[i]);
            dfs(ref result, candidates, target-candidates[i], ref combination, i);
            combination.RemoveAt(combination.Count - 1);
        }
    }
}
